File: src/workbench/growth.ts
import { arcTable, at, clamp, type Curve, type Point } from './geometry';
import type { Layout } from './model';
import { distance, lerp, polyPath, unionOutline } from './outlineUnion';
import {composedGrowth} from './composedGrowth';
import {spiralAnatomy} from './spiralAnatomy';

export interface GrowthSettings { seed:number; branches:number; reach:number; curl:number; levels:number; leaves:number; clearance:number; stem:number; side:'alternate'|'left'|'right'; family?:'spiral'|'spray'|'border'|'fan'|'branching'; composition?:1|2; secondaryScale?:number; sweeps?:1|2; construction?:'spiral'|'branching' }
export const defaultGrowth:GrowthSettings={seed:1248,branches:5,reach:33,curl:1,levels:2,leaves:2,clearance:2,stem:2.8,side:'alternate'};
export interface GrowthPart { id:string; parent:string|null; kind:'backbone'|'primary'|'secondary'|'leaf'; points:Point[]; polygon:Point[]; folds:Point[][]; length:number; width:number; birth:number; duration:number; contourSplit?:number }
export interface Trial {points:Point[];reason:'edge'|'spacing';parent:string}
export interface GrowthResult {parts:GrowthPart[];trials:Trial[];skipped:number;message:string;attempts:number}
function random(seed:number){let s=seed>>>0;return ()=>{s+=0x6D2B79F5;let t=Math.imul(s^s>>>15,s|1);t^=t+Math.imul(t^t>>>7,t|61);return ((t^t>>>14)>>>0)/4294967296;};}
export function lineLength(p:Point[]){return p.slice(1).reduce((sum,q,i)=>sum+distance(q,p[i]),0);}
export function lineFrame(points:Point[],progress:number){
  const target=clamp(progress,0,1)*lineLength(points);let total=0;
  for(let i=1;i<points.length;i++){const a=points[i-1],b=points[i],d=distance(a,b);if(total+d>=target||i===points.length-1){const t=clamp((target-total)/(d||1),0,1);return {point:lerp(a,b,t),angle:Math.atan2(b.y-a.y,b.x-a.x)};}total+=d;}
  return {point:points[0],angle:0};
}
export function ribbon(points:Point[],width:number):Point[]{
  const left:Point[]=[],right:Point[]=[];
  points.forEach((p,i)=>{
    const a=points[Math.max(0,i-1)],b=points[Math.min(points.length-1,i+1)],angle=Math.atan2(b.y-a.y,b.x-a.x);
    const half=width*(.06+.94*(1-i/(points.length-1))**.75)/2;
    left.push({x:p.x-Math.sin(angle)*half,y:p.y+Math.cos(angle)*half});right.push({x:p.x+Math.sin(angle)*half,y:p.y-Math.cos(angle)*half});
  });return [...left,...right.reverse()];
}
/** A growth tip advances along a tangent with increasing turning rate and
 * decaying step size. Its root tangent agrees with its parent. */
export function growCurl(root:Point,angle:number,reach:number,curl:number,side:number):Point[]{
  const points=[root],steps=140,decay=.24;
  // A logarithmic spiral has a shrinking radius and a known root tangent.
  // Align that tangent with the parent instead of integrating a loose hook.
  const rotation=angle-Math.atan2(side,-decay);
  for(let i=1;i<=steps;i++){
    const theta=i/steps*Math.PI*2*1.25*curl,r=reach*Math.exp(-decay*theta);
    const x=r*Math.cos(theta)-reach,y=side*r*Math.sin(theta);
    points.push({x:root.x+x*Math.cos(rotation)-y*Math.sin(rotation),y:root.y+x*Math.sin(rotation)+y*Math.cos(rotation)});
  }
  return points;
}
export function growSweep(root:Point,angle:number,reach:number,curl:number,side:number):Point[]{
  const map=(x:number,y:number)=>({x:root.x+x*Math.cos(angle)-y*Math.sin(angle),y:root.y+x*Math.sin(angle)+y*Math.cos(angle)});
  const end=map(reach*1.35,side*reach*.6),heading=angle+side*.8;
  const curve:Curve=[root,map(reach*.65,0),{x:end.x-Math.cos(heading)*reach*.4,y:end.y-Math.sin(heading)*reach*.4},end];
  return [...Array.from({length:81},(_,i)=>at(curve,i/80)),...growCurl(end,heading,reach*.55,curl,side).slice(1)];
}
export function generateGrowth(layout:Layout,settings:GrowthSettings=defaultGrowth):GrowthResult {
  if(layout.extraCurves?.length){
    const results=[layout.curve,...layout.extraCurves].map((curve,index)=>{const r=generateGrowth({...layout,curve,extraCurves:undefined,backboneGrowth:undefined,lockedParts:layout.lockedParts?.filter(p=>p.id.startsWith(`backbone-${index}/`)).map(p=>({...p,id:p.id.split('/').at(-1)!,parent:p.parent?p.parent.split('/').at(-1)!:null}))},layout.backboneGrowth?.[index]??settings);const id=(s:string)=>`backbone-${index}/${s}`;return {...r,parts:r.parts.map(p=>({...p,id:id(p.id),parent:p.parent?id(p.parent):null})),trials:r.trials.map(t=>({...t,parent:id(t.parent)}))};});
    return {parts:results.flatMap(r=>r.parts),trials:results.flatMap(r=>r.trials),attempts:0,skipped:0,message:`${results.length} backbones · independently grown scrolls`};
  }
  settings=layout.backboneGrowth?.[0]??settings;
  if(lineLength(arcTable(layout.curve).map(r=>r.point))<10)return {parts:[],trials:[],attempts:0,skipped:0,message:'Draw a longer guide to grow a scroll.'};
  // Construct at a consistent design scale so millimeter page dimensions do
  // not leave curls and leaf widths capped at their small-page sizes.
  const scale=Math.sqrt(layout.width*layout.height/(240*150));
  const shrink=(p:Point)=>({x:p.x/scale,y:p.y/scale});
  const normalized={...layout,lockedParts:layout.lockedParts?.map(p=>({...p,points:p.points.map(shrink),polygon:p.polygon.map(shrink),folds:p.folds.map(f=>f.map(shrink)),width:p.width/scale,length:p.length/scale})),width:layout.width/scale,height:layout.height/scale,curve:layout.curve.map(p=>({x:p.x/scale,y:p.y/scale})) as Curve};
  const result=(settings.composition===2||(settings.family&&settings.family!=='spiral'))?composedGrowth(normalized,settings):spiralAnatomy(normalized,settings);
  const resize=(p:Point)=>({x:p.x*scale,y:p.y*scale});
  return {...result,parts:result.parts.map(p=>layout.lockedParts?.find(q=>q.id===p.id)??({...p,points:p.points.map(resize),polygon:p.polygon.map(resize),folds:p.folds.map(f=>f.map(resize)),width:p.width*scale,length:p.length*scale})),trials:result.trials.map(t=>({...t,points:t.points.map(resize)}))};
}
export function growthDrawing(result:GrowthResult,progress=1){
  const visible=result.parts.filter(p=>progress>=p.birth+p.duration);
  const active=result.parts.filter(p=>progress>=p.birth&&progress<p.birth+p.duration);
  const polygons=visible.map(p=>p.polygon),folds=visible.flatMap(p=>p.folds);
  for(const p of active){const t=clamp((progress-p.birth)/p.duration,0,1);if(p.kind==='leaf'){
    if(t>.1){const root=p.points[0];polygons.push(p.polygon.map(q=>lerp(root,q,t)));}
  }else{const count=Math.max(2,Math.ceil(p.points.length*t));polygons.push(p.contourSplit?[...p.polygon.slice(0,Math.max(2,Math.ceil(p.contourSplit*t))),...p.polygon.slice(p.polygon.length-Math.max(2,Math.ceil((p.polygon.length-p.contourSplit)*t)))]:p.polygon.length===p.points.length*2?[...p.polygon.slice(0,count),...p.polygon.slice(p.polygon.length-count)]:ribbon(p.points.slice(0,count),p.width));}}
  return {outline:polygons.length?unionOutline(polygons):'',folds:folds.map(p=>polyPath(p)).join(' ')};
}
export function growthSvg(layout:Layout,settings:GrowthSettings){
  const drawing=growthDrawing(generateGrowth(layout,settings));
  return `<svg xmlns="http://www.w3.org/2000/svg" width="${layout.width}mm" height="${layout.height}mm" viewBox="0 0 ${layout.width} ${layout.height}"><title>Procedural scroll — seed ${settings.seed}</title><g fill="none" stroke="#000" stroke-linecap="round" stroke-linejoin="round"><path d="${drawing.outline}" stroke-width="0.35"/><path d="${drawing.folds}" stroke-width="0.2"/></g></svg>`;
}

File: core/src/growth.rs
//! Growth model: settings, grown parts, the curl primitive and the entry
//! point that grows one or more backbones (with hand-edited shoots).
use crate::geometry::{arc_table, clamp, guide_points, lerp, line_length, pt, Curve, Point};
use crate::shoots::{apply_shoot_edits, ShootEdit, ShootParams};
use std::f64::consts::PI;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Side { Alternate, Left, Right }
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Family { Spiral, Spray, Border, Fan, Branching }

#[derive(Clone, Debug, PartialEq)]
pub struct GrowthSettings {
    pub seed: u32, pub branches: f64, pub reach: f64, pub curl: f64, pub levels: u8, pub leaves: u8,
    pub clearance: f64, pub stem: f64, pub side: Side, pub family: Option<Family>, pub composition: Option<u8>,
    pub secondary_scale: Option<f64>, pub sweeps: Option<u8>, pub auto_shoots: Option<bool>, pub flip: Option<bool>,
    /// Hand-placed backbone: keep the scroll's proportions instead of
    /// shrinking it to fit inside the page.
    pub free: Option<bool>,
    /// Grows out of this other backbone: its start sits on that stem and the
    /// join merges like a leaf root.
    pub attach: Option<usize>,
    /// Wrapping leaves laid into the scroll's curl (0 to 2); the volute opens up to make room.
    pub wraps: Option<u8>,
    /// Library leaf (preset id) used for the wrapping leaves; None grows the plain acanthus wrap.
    pub wrap_leaf: Option<String>,
    /// A sheathing collar over the join, for a backbone grown from another:
    /// its size against the stems (1 = default); None is no collar.
    pub collar: Option<f64>,
    /// Collar style id ("axil" or "split"); None is the axil leaf.
    pub collar_style: Option<String>,
}
impl Default for GrowthSettings {
    fn default() -> Self { GrowthSettings { seed: 1248, branches: 5.0, reach: 33.0, curl: 1.0, levels: 2, leaves: 2, clearance: 2.0, stem: 2.8, side: Side::Alternate, family: None, composition: None, secondary_scale: None, sweeps: None, auto_shoots: None, flip: None, free: None, attach: None, wraps: None, wrap_leaf: None, collar: None, collar_style: None } }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Kind { Backbone, Primary, Secondary, Leaf }

#[derive(Clone, Debug)]
pub struct GrowthPart {
    pub id: String, pub parent: Option<String>, pub kind: Kind,
    pub points: Vec<Point>, pub polygon: Vec<Point>, pub folds: Vec<Vec<Point>>, pub ridges: Option<Vec<Vec<Point>>>,
    /// Inner edges drawn at outline weight (buds made of overlapping pieces).
    pub cuts: Vec<Vec<Point>>,
    pub length: f64, pub width: f64, pub birth: f64, pub duration: f64, pub contour_split: Option<usize>,
    pub shoot: Option<ShootParams>, pub under: bool,
}
#[derive(Clone, Debug, Default)]
pub struct GrowthResult { pub parts: Vec<GrowthPart>, pub skipped: usize, pub attempts: usize, pub message: String }

/// Mulberry32, bit-exact with the web version.
pub struct Mulberry(pub u32);
impl Mulberry {
    pub fn next(&mut self) -> f64 {
        self.0 = self.0.wrapping_add(0x6D2B79F5);
        let s = self.0;
        let mut t = (s ^ (s >> 15)).wrapping_mul(s | 1);
        t ^= t.wrapping_add((t ^ (t >> 7)).wrapping_mul(t | 61));
        (t ^ (t >> 14)) as f64 / 4294967296.0
    }
}

pub fn ribbon(points: &[Point], width: f64) -> Vec<Point> {
    let (mut left, mut right) = (vec![], vec![]);
    let n = points.len();
    for (i, p) in points.iter().enumerate() {
        let a = points[i.saturating_sub(1)]; let b = points[(i + 1).min(n - 1)]; let angle = (b.y - a.y).atan2(b.x - a.x);
        let half = width * (0.06 + 0.94 * (1.0 - i as f64 / (n as f64 - 1.0)).powf(0.75)) / 2.0;
        left.push(pt(p.x - angle.sin() * half, p.y + angle.cos() * half)); right.push(pt(p.x + angle.sin() * half, p.y - angle.cos() * half));
    }
    right.reverse(); left.extend(right); left
}
/// Logarithmic curl whose root tangent is `angle`; side ±1 picks the turn.
pub fn grow_curl(root: Point, angle: f64, reach: f64, curl: f64, side: f64) -> Vec<Point> {
    let (steps, decay) = (140, 0.24);
    let rotation = angle - side.atan2(-decay);
    let mut points = vec![root];
    for i in 1..=steps {
        let theta = i as f64 / steps as f64 * PI * 2.0 * 1.25 * curl; let r = reach * (-decay * theta).exp();
        let x = r * theta.cos() - reach; let y = side * r * theta.sin();
        points.push(pt(root.x + x * rotation.cos() - y * rotation.sin(), root.y + x * rotation.sin() + y * rotation.cos()));
    }
    points
}

/// Everything a backbone needs to grow: page, curve, locks, edits, settings.
pub struct GrowInput<'a> {
    pub width: f64, pub height: f64, pub curve: Curve,
    pub locked: &'a [GrowthPart], pub shoots: &'a [ShootEdit], pub settings: &'a GrowthSettings,
}

/// Grow one backbone in page units.
pub fn grow_backbone(inp: &GrowInput) -> GrowthResult {
    let s = inp.settings;
    if line_length(&guide_points(&inp.curve)) < 10.0 { return GrowthResult { message: "Draw a longer guide to grow a scroll.".into(), ..Default::default() }; }
    // Grow at one design scale so page size does not cap curls and widths.
    let scale = (inp.width * inp.height / (240.0 * 150.0)).sqrt();
    let shrink = |p: &Point| pt(p.x / scale, p.y / scale);
    let locked: Vec<GrowthPart> = inp.locked.iter().map(|p| GrowthPart { points: p.points.iter().map(shrink).collect(), polygon: p.polygon.iter().map(shrink).collect(), folds: p.folds.iter().map(|f| f.iter().map(shrink).collect()).collect(), ridges: p.ridges.as_ref().map(|r| r.iter().map(|f| f.iter().map(shrink).collect()).collect()), cuts: p.cuts.iter().map(|f| f.iter().map(shrink).collect()).collect(), width: p.width / scale, length: p.length / scale, ..p.clone() }).collect();
    let page = Page { width: inp.width / scale, height: inp.height / scale, curve: inp.curve.map(|p| shrink(&p)), locked: &locked };
    let grown = if s.composition == Some(2) || s.family.map_or(false, |f| f != Family::Spiral) { crate::composed::composed_growth(&page, s) } else { crate::spiral::spiral_anatomy(&page, s) };
    let own = if s.auto_shoots == Some(false) { GrowthResult { parts: grown.parts.into_iter().filter(|p| p.parent.is_none() || locked.iter().any(|q| q.id == p.id)).collect(), ..grown } } else { grown };
    // Wrapping leaves follow the main scroll into its eye, above it and
    // beneath any leaves placed afterwards.
    let own = match s.wraps.filter(|&n| n > 0) {
        Some(n) => { let mut parts = own.parts; if let Some(i) = parts.iter().position(|p| p.parent.is_none()) { let w = crate::wraps::wrapping_leaves(&parts[i], n, s.leaves > 0, s.wrap_leaf.as_deref().and_then(crate::profiles::profile)); parts.splice(i + 1..i + 1, w); } GrowthResult { parts, ..own } }
        None => own,
    };
    let result = apply_shoot_edits(own, &guide_points(&page.curve), inp.shoots, s.leaves > 0);
    let resize = |p: &Point| pt(p.x * scale, p.y * scale);
    let parts = result.parts.into_iter().map(|p| {
        if let Some(l) = inp.locked.iter().find(|q| q.id == p.id) { return l.clone(); }
        GrowthPart { points: p.points.iter().map(resize).collect(), polygon: p.polygon.iter().map(resize).collect(), folds: p.folds.iter().map(|f| f.iter().map(resize).collect()).collect(), ridges: p.ridges.as_ref().map(|r| r.iter().map(|f| f.iter().map(resize).collect()).collect()), cuts: p.cuts.iter().map(|f| f.iter().map(resize).collect()).collect(), width: p.width * scale, length: p.length * scale, ..p }
    }).collect();
    GrowthResult { parts, ..result }
}

/// The normalised page a family grows on.
pub struct Page<'a> { pub width: f64, pub height: f64, pub curve: Curve, pub locked: &'a [GrowthPart] }
impl Page<'_> { pub fn guide(&self) -> Vec<Point> { arc_table(&self.curve).into_iter().map(|r| r.point).collect() } }

/// Simplified outline during the growth animation (progress < 1).
pub fn growth_polygons(result: &GrowthResult, progress: f64) -> (Vec<Vec<Point>>, Vec<Vec<Point>>) {
    let mut polygons = vec![]; let mut folds = vec![];
    for p in &result.parts {
        if progress >= p.birth + p.duration { polygons.push(p.polygon.clone()); folds.extend(p.folds.iter().cloned()); continue; }
        if progress < p.birth { continue; }
        let t = clamp((progress - p.birth) / p.duration, 0.0, 1.0);
        if p.kind == Kind::Leaf { if t > 0.1 { let root = p.points[0]; polygons.push(p.polygon.iter().map(|q| lerp(root, *q, t)).collect()); } continue; }
        let count = 2usize.max((p.points.len() as f64 * t).ceil() as usize);
        if let Some(split) = p.contour_split {
            let a = 2usize.max((split as f64 * t).ceil() as usize); let b = 2usize.max(((p.polygon.len() - split) as f64 * t).ceil() as usize);
            let mut poly = p.polygon[..a.min(p.polygon.len())].to_vec(); poly.extend_from_slice(&p.polygon[p.polygon.len().saturating_sub(b)..]); polygons.push(poly);
        } else if p.polygon.len() == p.points.len() * 2 {
            let mut poly = p.polygon[..count].to_vec(); poly.extend_from_slice(&p.polygon[p.polygon.len() - count..]); polygons.push(poly);
        } else { polygons.push(ribbon(&p.points[..count.min(p.points.len())], p.width)); }
    }
    (polygons, folds)
}
